### PR TITLE
Use single mutation during no chop fence upgrade

### PR DESCRIPTION
**Pull request by EdColeman**
This is an optimization of the upgrade code from PR #3876.  This code processes each row and emits one mutation for all of the changes needed to upgrade that row.  At a minimum this should reduce the number of mutations needed by 50% by combining the `add updated file reference with fencing` and the `delete obsolete file reference` into a single mutation.  When there are multiple files, chopped or external compaction references for the row, the reduction will be greater. 

The changes simplified error handling so individual methods for each column type was not providing clarity and were removed.  The tests were modified using a mock scanner to provide equivalent coverage that was done for the individual methods.

This code was also testing manually upgrading an uno instance from 2.1.3-SNAPSHOT to this code.

### CODE REVIEW

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java [hunk 1]
```diff
@@ -84,111 +89,97 @@ public void upgradeZookeeper(@NonNull ServerContext context) {
...
         } else {
           throw new IllegalStateException("Processing: " + tableName
               + " Received unexpected column family processing references: " + family);
         }
-      });
+      }
+      // send last mutation
+      if (update != null) {
+        log.trace("table: {}, update: {}", tableName, update.prettyPrint());
+        batchWriter.addMutation(update);
```
Comment (cshannon) [inline server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:166]:
```suggestion
        if (!update.getUpdates().isEmpty()) {
          batchWriter.addMutation(update);
        }
```

Same as other comment

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java [hunk 2]
```diff
@@ -84,111 +89,97 @@ public void upgradeZookeeper(@NonNull ServerContext context) {
...
+        Key key = entry.getKey();
+        Value value = entry.getValue();
+
+        // on new row, write current mutation and prepare a new one.
+        Text r = key.getRow();
+        if (update == null) {
+          update = new Mutation(r);
+        } else if (!Arrays.equals(update.getRow(), r.getBytes())) {
+          log.trace("table: {}, update: {}", tableName, update.prettyPrint());
+          batchWriter.addMutation(update);
```
Comment (cshannon) [inline server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:139]:
```suggestion
        if (!update.getUpdates().isEmpty()) {
          batchWriter.addMutation(update);
        }
```

I'm wondering if we need to check if the mutation is empty here before adding it. I don't _think_ this necessary the way things are currently written as we are only fetching 3 columns and each of those should cause a mutation to be added but it might not hurt to check since adding an empty mutation would throw an exception.

**Review by cshannon — COMMENTED** [review]
I think it would be good to add a quick test to make sure there's no errors for tablets that don't have the 3 columns that are being upgraded (it just skips over it).

As I said in my in line comments, we could add a check for an empty mutation but you might not need to worry about it 

Otherwise things look fine to me

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java (hunk 1 above)
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:168]: Guard added in 76b099c584

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java (hunk 2 above)
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java]: Guard added in 76b099c584

Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java]: Also added test to validate batch write not called when the mutation is empty,

**Review by cshannon — APPROVED** [review]

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java [hunk 3]
```diff
@@ -84,111 +89,99 @@ public void upgradeZookeeper(@NonNull ServerContext context) {
...
+      Mutation update = null;
+      for (Map.Entry<Key,Value> entry : scanner) {
+        Key key = entry.getKey();
+        Value value = entry.getValue();
+
+        // on new row, write current mutation and prepare a new one.
+        Text r = key.getRow();
+        if (update == null) {
+          update = new Mutation(r);
+        } else if (!Arrays.equals(update.getRow(), r.getBytes())) {
```
Comment (keith-turner) [inline server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:137]: Text.getBytes() is dangerous.  It may return an array that is longer than the data in the Text object.  Can use the Accumulo TextUtil class to deal with this.  Or could call Text.copyBytes()

```suggestion
        } else if (!Arrays.equals(update.getRow(), TextUtil.getBytes(r.getBytes()))) {
```

Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java]: Addressed in b1c89e3ce4

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java [hunk 4]
```diff
@@ -84,111 +91,100 @@ public void upgradeZookeeper(@NonNull ServerContext context) {
...
+        Key key = entry.getKey();
+        Value value = entry.getValue();
+        Preconditions.checkState(key.getColumnVisibilityData().length() == 0,
+            "Expected empty visibility, saw %s ", key.getColumnVisibilityData());
+        // on new row, write current mutation and prepare a new one.
+        Text r = key.getRow();
+        if (update == null) {
+          update = new Mutation(r);
+        } else if (!Arrays.equals(update.getRow(), TextUtil.getBytes(r))) {
+          log.trace("table: {}, update: {}", tableName, update.prettyPrint());
```
Comment (keith-turner) [inline server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:141]: May want to wrap this in `if(is trace enabled)` because `update.prettyPrint()` function will do work to compute a string.

File: server/manager/src/test/java/org/apache/accumulo/manager/upgrade/Upgrader11to12Test.java [hunk 5]
```diff
@@ -62,198 +67,334 @@ public class Upgrader11to12Test {
...
 
-    upgrader.upgradeDataFileCF(k, v, bw, "aTable");
+    Mutation upgrade = new Mutation(k.getRow());
+    upgrader.upgradeDataFileCF(k, v, upgrade);
 
-    StoredTabletFile stf = StoredTabletFile.of(new Path(fileName));
-    Mutation add = new Mutation(k.getRow()).at().family(DataFileColumnFamily.NAME)
-        .qualifier(stf.getMetadataText()).put(v);
-    LOG.debug("add mutation to be expected: {}", add.prettyPrint());
+    var pending = upgrade.getUpdates();
```
Comment (keith-turner) [inline server/manager/src/test/java/org/apache/accumulo/manager/upgrade/Upgrader11to12Test.java:81]: could check the the size of pending is 2

File: server/manager/src/test/java/org/apache/accumulo/manager/upgrade/Upgrader11to12Test.java [hunk 6]
```diff
@@ -62,198 +67,334 @@ public class Upgrader11to12Test {
...
+    LOG.info("c:{}", capturedUpdate1.getValue().prettyPrint());
+    var u1 = capturedUpdate1.getValue();
+    // 2 file add, 2 file delete. 1 chop delete, 1 ext comp delete
+    assertEquals(6, u1.getUpdates().size());
 
-    replay(bw);
+    LOG.info("c:{}", capturedUpdate2.getValue().prettyPrint());
+    var u2 = capturedUpdate2.getValue();
+    // 1 add, 1 delete
+    assertEquals(2, u2.getUpdates().size());
```
Comment (keith-turner) [inline server/manager/src/test/java/org/apache/accumulo/manager/upgrade/Upgrader11to12Test.java:169]: could also stream over the updates and filter deletes and then count and assert that count is 1
```suggestion
    assertEquals(2, u2.getUpdates().size());
    assertEquals(1, u2.getUpdates().stream().filter(cup->cup.isDeleted()).count());
```

File: server/manager/src/test/java/org/apache/accumulo/manager/upgrade/Upgrader11to12Test.java [hunk 7]
```diff
@@ -62,198 +67,334 @@ public class Upgrader11to12Test {
...
-    Value v = new Value("1234,5678");
+    Upgrader11to12 upgrader = new Upgrader11to12();
+    upgrader.processReferences(batchWriter, scanner, "accumulo.metadata");
 
-    assertThrows(IllegalArgumentException.class,
-        () -> upgrader.upgradeDataFileCF(k, v, bw, "aTable"));
+    LOG.info("c:{}", capturedUpdate1.getValue().prettyPrint());
+    var u1 = capturedUpdate1.getValue();
+    // 1 add, 1 delete
+    assertEquals(2, u1.getUpdates().size());
```
Comment (keith-turner) [inline server/manager/src/test/java/org/apache/accumulo/manager/upgrade/Upgrader11to12Test.java:216]: could also stream and count deletes here

**Review by keith-turner — APPROVED** [review]
Nice tests.  Curious if you have run code coverage on the test, seems like they would get good code coverage.

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java [hunk 8]
```diff
@@ -84,111 +91,102 @@ public void upgradeZookeeper(@NonNull ServerContext context) {
   public void upgradeRoot(@NonNull ServerContext context) {
     log.debug("Upgrade root: upgrading to data version {}", METADATA_FILE_JSON_ENCODING);
     var rootName = Ample.DataLevel.METADATA.metaTable();
-    processReferences(context, rootName);
+    // not using ample to avoid StoredTabletFile because old file ref is incompatible
+    try (AccumuloClient c = Accumulo.newClient().from(context.getProperties()).build();
+        BatchWriter batchWriter = c.createBatchWriter(rootName); Scanner scanner =
```
Comment (ctubbsii) [inline server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:96]: context is already an an instance of AccumuloClient. You don't need to create a new one. I see this is an issue with the existing upgrader code prior to this PR, but it should still be fixed. Funny thing is that the scanner is already created this way.

```suggestion
    try (var batchWriter = context.createBatchWriter(rootName); var scanner =
```

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java [hunk 9]
```diff
@@ -84,111 +91,102 @@ public void upgradeZookeeper(@NonNull ServerContext context) {
...
+    }
   }
 
   @Override
   public void upgradeMetadata(@NonNull ServerContext context) {
     log.debug("Upgrade metadata: upgrading to data version {}", METADATA_FILE_JSON_ENCODING);
     var metaName = Ample.DataLevel.USER.metaTable();
-    processReferences(context, metaName);
+    try (AccumuloClient c = Accumulo.newClient().from(context.getProperties()).build();
+        BatchWriter batchWriter = c.createBatchWriter(metaName); Scanner scanner =
```
Comment (ctubbsii) [inline server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:113]:
```suggestion
    try (var batchWriter = context.createBatchWriter(metaName); var scanner =
```

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java (hunk 8 above)
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java]: Fixed in fe6e9e4172

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java (hunk 9 above)
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java]: Fixed in fe6e9e4172

File: server/manager/src/test/java/org/apache/accumulo/manager/upgrade/Upgrader11to12Test.java (hunk 7 above)
Comment (EdColeman) [reply server/manager/src/test/java/org/apache/accumulo/manager/upgrade/Upgrader11to12Test.java:217]: Added with e09da28032

File: server/manager/src/test/java/org/apache/accumulo/manager/upgrade/Upgrader11to12Test.java (hunk 6 above)
Comment (EdColeman) [reply server/manager/src/test/java/org/apache/accumulo/manager/upgrade/Upgrader11to12Test.java:169]: Added with e09da28032

File: server/manager/src/test/java/org/apache/accumulo/manager/upgrade/Upgrader11to12Test.java (hunk 5 above)
Comment (EdColeman) [reply server/manager/src/test/java/org/apache/accumulo/manager/upgrade/Upgrader11to12Test.java:81]: Added  with e09da28032

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java (hunk 4 above)
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java]: Changed in e09da28032
